test/integration/certificates/util.go

```diff
@@ -0,0 +1,173 @@
+		c, err := cmCl.CertmanagerV1().Certificates(cert.Namespace).Get(ctx, cert.Name, metav1.GetOptions{})
+		if err != nil {
+			// certificate not found
+			return false, err
+		}
+		if !apiutil.CertificateHasCondition(c, condition) {
+			t.Logf("Certificate does not have the expected %v conditon: %v, retrying", condition.Type, condition.Status)
+			return false, nil
+		}
+		//condition has been applied, run addition assertions
+		for _, a := range assertions {
+			a(t, c)
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Logf("expected Certificate to have %v condition set to %v", condition.Type, condition.Status)
+		t.Fatal(err)
+	}
+}
+
+type assertFunc func(*testing.T, *cmapi.Certificate)
+
+// mustGenerateCSRImpl returns PEM encoded certificate signing request.
+func mustGenerateCSRImpl(t *testing.T, pkData []byte, cert *cmapi.Certificate) []byte {
+	csrPEM, err := generateCSRImpl(cert, pkData)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return csrPEM
+}
+
+// mustCreatePEMPrivateKey returns a PEM encoded 2048 bit RSA private key.
+func mustCreatePEMPrivateKey(t *testing.T) []byte {
+	pk, err := pki.GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Fatal(err)
+	}
+	pkData, err := pki.EncodePrivateKey(pk, cmapi.PKCS8)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return pkData
+}
+
+// mustCreateCertWithNotBeforeAfter returns a self-signed X509 cert for Certificate
+// with the provided NotBefore, NotAfter values
+func mustCreateCertWithNotBeforeAfter(t *testing.T, pkData []byte, spec *cmapi.Certificate, notBefore, notAfter time.Time) []byte {
+	pk, err := pki.DecodePrivateKeyBytes(pkData)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	template, err := pki.GenerateTemplate(spec)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	template.NotBefore = notBefore
+	template.NotAfter = notAfter
+
+	certData, _, err := pki.SignCertificate(template, template, pk.Public(), pk)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return certData
+}
+
+// mustCreateCert returns a self-signed X509 certificate
+func mustCreateCert(t *testing.T, pkData []byte, spec *cmapi.Certificate) []byte {
+	pk, err := pki.DecodePrivateKeyBytes(pkData)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	template, err := pki.GenerateTemplate(spec)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	certData, _, err := pki.SignCertificate(template, template, pk.Public(), pk)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return certData
+}
+
+// generateCSRImpl returns a PEM encoded certificate signing request for the certificate
+func generateCSRImpl(crt *cmapi.Certificate, pk []byte) ([]byte, error) {
+	csr, err := pki.GenerateCSR(crt)
+	if err != nil {
+		return nil, err
+	}
+
+	signer, err := pki.DecodePrivateKeyBytes(pk)
+	if err != nil {
+		return nil, err
+	}
+
+	csrDER, err := pki.EncodeCSR(csr, signer)
+	if err != nil {
+		return nil, err
+	}
+
+	csrPEM := pem.EncodeToMemory(&pem.Block{
+		Type: "CERTIFICATE REQUEST", Bytes: csrDER,
+	})
+
+	return csrPEM, nil
```

Comment:
Is there not a more generic place where this can go? There seems to be a lot of duplication with this kind of code though out the code base.

Reply:
I was thinking about that and couldn't decide if I should do that as a part of this PR or whether we need some bigger restructure of all these kind of util function. I'll look into putting them in a generic location with this PR.